[GraphQL] Add custom queries

lukasluecke opened the pull request:
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | 
| License       | MIT
| Doc PR        | TODO

Possibly related: #1700 #473 #2309 #1870

Currently partially implements custom queries on existing resources, but the goal would be at least full custom query and mutation support on existing resources, possibly even for "new" resources, custom types etc. (maybe in another PR)

**Usage example**

`User.php` (Resource class)
```php
/**
 * @ApiResource(graphql={
 *     "query",
 *     "delete",
 *     "update",
 *     "create",
 *     "current"={
 *          "item"=CurrentUserResolver::class,
 *          "description"="Get currently logged in user"
 *     },
 *     "active"={
 *          "collection"=ActiveUsersResolver::class,
 *          "description"="Get all active users"
 *     }
 * })
 */
```

`CurrentUserResolver.php`
```php
class CurrentUserResolver implements QueryResolverInterface
{
    private $tokenStorage;
    private $normalizer;

    public function __construct(TokenStorageInterface $tokenStorage, NormalizerInterface $normalizer)
    {
        $this->tokenStorage = $tokenStorage;
        $this->normalizer = $normalizer;
    }

    public function __invoke($source, $args, $context, ResolveInfo $info)
    {
        $token = $this->tokenStorage->getToken();

        if (!$token) return null;

        return $this->normalizer->normalize($token->getUser(), ItemNormalizer::FORMAT);
    }
}
```

`ActiveUsersResolver.php`
```php
class ActiveUsersResolver implements QueryResolverInterface
{
    private $repository;
    private $normalizer;

    public function __construct(UserRepository $repository, NormalizerInterface $normalizer)
    {
        $this->repository = $repository;
        $this->normalizer = $normalizer;
    }

    public function __invoke($source, $args, $context, ResolveInfo $info)
    {
        $collection = $this->repository->findActive();

        $offset = 0;
        $data = ['totalCount' => 0.0, 'edges' => [], 'pageInfo' => ['endCursor' => null, 'hasNextPage' => false]];

        foreach ($collection as $index => $object) {
            $data['edges'][$index] = [
                'node' => $this->normalizer->normalize($object, ItemNormalizer::FORMAT),
                'cursor' => base64_encode((string) ($index + $offset)),
            ];
        }

        return $data;
    }
}
```

lukasluecke commented:
@alanpoulain Any early input from you, while I continue working on this? (e.g. the configuration structure, scope of this feature, ..)

alanpoulain commented:
I will have a look this weekend. @raoulclais is working on the custom mutations, he will probably make a PR soon.

lukasluecke commented:
@alanpoulain Thank you

@raoulclais Is there anywhere I can check out your current progress on the custom mutations?

alanpoulain commented on src/Bridge/Symfony/Bundle/DependencyInjection/Compiler/GraphQlResolverPass.php:
```diff
@@ -0,0 +1,50 @@
+ *
+ * @author Lukas Lücke <lukas@luecke.me>
+ */
+final class GraphQlResolverPass implements CompilerPassInterface
```
`GraphQlQueryResolverPass` instead?

alanpoulain commented on src/GraphQl/Resolver/ResolverInterface.php:
```diff
@@ -0,0 +1,35 @@
+ *
+ * @author Lukas Lücke <lukas@luecke.me>
+ */
+interface ResolverInterface
```
`QueryResolverInterface` instead?

alanpoulain commented:
I think it should be better to have something like this:
```php

/**
 * @ApiResource(graphql={
 *     "query",
 *     "delete",
 *     "update",
 *     "create",
 *    "myQuery"={
 *        "query"=MyQuery::class
 *    }
 * })
 */
```
Like the PR for the custom mutations: https://github.com/api-platform/core/pull/2447
You should be able to identify them with the `query` key.

alanpoulain commented on src/GraphQl/Type/SchemaBuilder.php:
```diff
@@ -172,6 +183,38 @@ private function getQueryFields(string $resourceClass, ResourceMetadata $resourc
+    /**
+     * Gets additional custom query fields of the schema.
+     */
+    private function getCustomQueryFields(string $resourceClass, ResourceMetadata $resourceMetadata): array
```
I think it should be possible to reuse the `getQueryFields` method.

lukasluecke commented:
@alanpoulain Thank you for your feedback. I have incorporated your suggestions and updated the resulting code sample in the PR description.

I'm not really sure about the configuration still. We need more than just a `query` field, as suggested by you. For now I have chosen `item` and `collection` fields, which serve both as the resolver class, and also as an indicator for what kind of query it is.

But if we go further, e.g custom input-types or result types for collections I'm not sure this will hold up.

alanpoulain commented:
IMO it should not be an issue in the future.
We would have something like this:
```php
/**
 * @ApiResource(graphql={
 *     "query",
 *     "delete",
 *     "update",
 *     "create",
 *     "current"={
 *          "item"=CurrentUserResolver::class,
 *          "description"="Get currently logged in user",
 *          "input"={"type"=MyType::class, "description"="Custom description input", args=[]}
 *     },
 *     "active"={
 *          "collection"=ActiveUsersResolver::class,
 *          "description"="Get all active users"
 *     }
 * })
 */
```
WDYT?

I think we need to add the "query" term anyway. My suggestion: `item_query` and `collection_query`. Are you OK with this?

alanpoulain commented on src/Bridge/Symfony/Bundle/DependencyInjection/Compiler/GraphQlQueryResolverPass.php:
```diff
@@ -0,0 +1,50 @@
+    /**
+     * {@inheritdoc}
+     *
+     * @throws RuntimeException
```
Can be removed.

lukasluecke commented:
LGTM. I‘ll make the changes sometime next week, probably towards the end (currently on vacation) 👍🏻

> Am 02.02.2019 um 19:20 schrieb Alan Poulain <notifications@github.com>:
> 
> IMO it should not be an issue in the future.
> We would have something like this:
> 
> /**
>  * @ApiResource(graphql={
>  *     "query",
>  *     "delete",
>  *     "update",
>  *     "create",
>  *     "current"={
>  *          "item"=CurrentUserResolver::class,
>  *          "description"="Get currently logged in user",
>  *          "input"={"type"=MyType::class, "description"="Custom description input", args=[]}
>  *     },
>  *     "active"={
>  *          "collection"=ActiveUsersResolver::class,
>  *          "description"="Get all active users"
>  *     }
>  * })
>  */
> WDYT?
> 
> I think we need to add the "query" term anyway. My suggestion: item_query and collection_query. Are you OK with this?
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.

SCIF commented:
Guys, don't you think that custom operation controller/invoked class should return the type of resource/iterable collection of this resource objects like it's done for REST?

However, the resource term doesn't fit very good for graphql and it's more about operations and types. What do you think about idea to introduce `ApiOperation` annotation providing easy way to handle custom operations? Something like:

```
/**
 * @ApiOperation(iri="somecustomname", arguments={…)
 */
final class AddSubscription
{
    public function __constructor(…)
    {
        …
    }

    public function __invoke(…): Subscription
    {
        …
    } 
}
```

Returning type could easily detected by analysing `__invoke()` method.

My apologies in case if I'm wrong. it's just a suggestion.

lukasluecke commented:
@alanpoulain I have updated the config keys as agreed, and think this could be merged now. Further improvements (e.g. custom inputs) will follow in separate PRs.

(Sorry for the delay, had the flu after my vacation 🙄)

lukasluecke commented:
@SCIF Thanks fo your input and suggestion. I am not opposed to supporting annotations for this, but think it could be added later on as an additional / alternative way of configuring this behaviour.

alanpoulain commented on src/Bridge/Symfony/Bundle/ApiPlatformBundle.php:
```diff
@@ -38,6 +39,7 @@ public function build(ContainerBuilder $container)
         $container->addCompilerPass(new DataProviderPass());
         $container->addCompilerPass(new AnnotationFilterPass());
         $container->addCompilerPass(new FilterPass());
+        $container->addCompilerPass(new GraphQlQueryResolverPass());
```
Maybe move it with the `GraphQlTypePass`?

alanpoulain commented on src/Bridge/Symfony/Bundle/DependencyInjection/Compiler/GraphQlQueryResolverPass.php:
```diff
@@ -0,0 +1,47 @@
+        $resolvers = [];
+        foreach ($container->findTaggedServiceIds('api_platform.graphql.query_resolver', true) as $serviceId => $tags) {
+            foreach ($tags as $tag) {
+                if (!isset($tag['id'])) {
```
```php
$types[$tag['id'] ?? $serviceId] = new Reference($serviceId);
```
Like the `GraphQlTypePass`.

lukasluecke replied:
I originally based it on the FilterPass, which used the "old" syntax.

alanpoulain commented on src/Bridge/Symfony/Bundle/Resources/config/graphql.xml:
```diff
@@ -65,9 +65,14 @@
             <argument>%api_platform.collection.pagination.enabled%</argument>
         </service>
 
+        <service id="api_platform.graphql.query_resolver_locator" class="Symfony\Component\DependencyInjection\ServiceLocator">
```
Add it with the resolvers?

lukasluecke replied:
Then it would be before the `$typesFactory` argument, which is not nullable right now. But the `$queryResolverLocator` is. Should I just make it required? (what was the consesus on the null / required discussion?)

alanpoulain replied:
That's not what I meant :slightly_smiling_face:
I was just saying the service definition should be with the other resolver service definitions (below `<!-- Resolvers -->`).
But to answer your question, since we decided that the `typesFactory` is required, there is already a breaking change. Therefore I think you can make it required and move it before the `typesFactory`.

lukasluecke replied:
Ah I see it now, makes sense. I'll make both changes 👍

alanpoulain commented on src/GraphQl/Resolver/QueryResolverInterface.php:
```diff
@@ -0,0 +1,29 @@
+ */
+interface QueryResolverInterface
+{
+    public function __invoke($source, $args, $context, ResolveInfo $info);
```
Maybe add:
```php
    /**
     * @return mixed The queried item
     */
```

alanpoulain commented on src/GraphQl/Type/SchemaBuilder.php:
```diff
@@ -86,12 +88,28 @@ public function getSchema(): Schema
             foreach ($graphqlConfiguration as $operationName => $value) {
                 if ('query' === $operationName) {
-                    $queryFields += $this->getQueryFields($resourceClass, $resourceMetadata);
+                    $queryFields += $this->getQueryFields($resourceClass, $resourceMetadata, $operationName, ['args' => ['id' => ['type' => GraphQLType::id()]]], true);
```
Why `true` for the `$collection` parameter?

alanpoulain replied:
Maybe an empty array will be better IMO.

lukasluecke replied:
Because that is the default behaviour. `query` generates both an item and a collection query (using the default input object).

lukasluecke replied:
Your choice. Would just need to modify the condition inside `getQueryFields` to explicitly check for `!== false` instead of "truthy" ;)

alanpoulain replied:
I think it would be better since it would also remove the `is_array` checks :wink:

lukasluecke replied:
Makes sense 👍

alanpoulain commented on src/GraphQl/Type/SchemaBuilder.php:
```diff
@@ -156,19 +174,26 @@ private function getNodeQueryField(): array
      * Gets the query fields of the schema.
      */
-    private function getQueryFields(string $resourceClass, ResourceMetadata $resourceMetadata): array
+    private function getQueryFields(string $resourceClass, ResourceMetadata $resourceMetadata, string $operationName = 'query', $item, $collection): array
```
```suggestion
    private function getQueryFields(string $resourceClass, ResourceMetadata $resourceMetadata, string $operationName, $itemConfiguration, $collectionConfiguration): array
```
I think you should add PHPDoc for the `$itemConfiguration` and `$collectionConfiguration` parameters like this:
```php
/**
 * @param array|bool $itemConfiguration false if not configured
 * @param array|bool $collectionConfiguration false if not configured
 */
```

alanpoulain commented on src/GraphQl/Type/SchemaBuilder.php:
```diff
@@ -155,20 +173,30 @@ private function getNodeQueryField(): array
+        $deprecationReason = $resourceMetadata->getGraphqlAttribute($operationName, 'deprecation_reason', '', true);
+
+        if ($itemConfiguration && $fieldConfiguration = $this->getResourceFieldConfiguration($resourceClass, $resourceMetadata, null, $deprecationReason, new Type(Type::BUILTIN_TYPE_OBJECT, true, $resourceClass), $resourceClass)) {
+            if (\is_array($itemConfiguration)) {
```
Should be not necessary if you use an empty array (see my previous comment).

alanpoulain commented on src/GraphQl/Type/SchemaBuilder.php:
```diff
@@ -155,20 +173,30 @@ private function getNodeQueryField(): array
-        if ($fieldConfiguration = $this->getResourceFieldConfiguration($resourceClass, $resourceMetadata, null, $deprecationReason, new Type(Type::BUILTIN_TYPE_OBJECT, false, null, true, null, new Type(Type::BUILTIN_TYPE_OBJECT, false, $resourceClass)), $resourceClass)) {
-            $queryFields[lcfirst(Inflector::pluralize($shortName))] = $fieldConfiguration;
+        if ($collectionConfiguration && $fieldConfiguration = $this->getResourceFieldConfiguration($resourceClass, $resourceMetadata, null, $deprecationReason, new Type(Type::BUILTIN_TYPE_OBJECT, false, null, true, null, new Type(Type::BUILTIN_TYPE_OBJECT, false, $resourceClass)), $resourceClass)) {
+            if (\is_array($collectionConfiguration)) {
```
Same.

alanpoulain commented on src/GraphQl/Type/SchemaBuilder.php:
```diff
@@ -155,20 +173,30 @@ private function getNodeQueryField(): array
+     * @param array|bool $collectionConfiguration false if not configured
      */
-    private function getQueryFields(string $resourceClass, ResourceMetadata $resourceMetadata): array
+    private function getQueryFields(string $resourceClass, ResourceMetadata $resourceMetadata, string $operationName = 'query', $itemConfiguration, $collectionConfiguration): array
```
```suggestion
    private function getQueryFields(string $resourceClass, ResourceMetadata $resourceMetadata, string $operationName, $itemConfiguration, $collectionConfiguration): array
```

alanpoulain commented on src/GraphQl/Type/SchemaBuilder.php:
```diff
@@ -86,12 +88,28 @@ public function getSchema(): Schema
                 }
 
-                $mutationFields[$operationName.$resourceMetadata->getShortName()] = $this->getMutationFields($resourceClass, $resourceMetadata, $operationName);
+                if ($item = $resourceMetadata->getGraphqlAttribute($operationName, 'item_query')) {
```
```suggestion
                if ($itemConfiguration = $resourceMetadata->getGraphqlAttribute($operationName, 'item_query')) {
```

lukasluecke replied:
`item_query` is not the same as `$itemConfiguration`

alanpoulain commented on src/GraphQl/Type/SchemaBuilder.php:
```diff
@@ -86,12 +88,28 @@ public function getSchema(): Schema
+                    continue;
+                }
+
+                if ($collection = $resourceMetadata->getGraphqlAttribute($operationName, 'collection_query')) {
```
```suggestion
                if ($collectionConfiguration = $resourceMetadata->getGraphqlAttribute($operationName, 'collection_query')) {
```

lukasluecke replied:
`collection_query` is not the same as `$collectionConfiguration`

alanpoulain commented on src/GraphQl/Type/SchemaBuilder.php:
```diff
@@ -86,12 +88,28 @@ public function getSchema(): Schema
 
-                $mutationFields[$operationName.$resourceMetadata->getShortName()] = $this->getMutationFields($resourceClass, $resourceMetadata, $operationName);
+                if ($item = $resourceMetadata->getGraphqlAttribute($operationName, 'item_query')) {
+                    $value['resolve'] = $this->queryResolverLocator->get($item);
```
```suggestion
                    $itemConfiguration['resolve'] = $this->queryResolverLocator->get($item);
```

lukasluecke replied:
`$value` is the whole configuration of the operation, whereas `$itemQuery` is only the FQDN of the resolver class.

alanpoulain commented on src/GraphQl/Type/SchemaBuilder.php:
```diff
@@ -86,12 +88,28 @@ public function getSchema(): Schema
+                }
+
+                if ($collection = $resourceMetadata->getGraphqlAttribute($operationName, 'collection_query')) {
+                    $value['resolve'] = $this->queryResolverLocator->get($collection);
```
```suggestion
                    $collectionConfiguration['resolve'] = $this->queryResolverLocator->get($collection);
```

lukasluecke replied:
`$value` is the whole configuration of the operation, whereas `$collectionQuery` is only the FQDN of the resolver class.

alanpoulain commented:
I think you need to add Behat tests and unit tests if possible to make sure it's working correctly.
Thank you for working on this, it will be a great improvement!

alanpoulain commented on src/GraphQl/Type/SchemaBuilder.php:
```diff
@@ -155,20 +173,30 @@ private function getNodeQueryField(): array
+
+        if ($itemConfiguration && $fieldConfiguration = $this->getResourceFieldConfiguration($resourceClass, $resourceMetadata, null, $deprecationReason, new Type(Type::BUILTIN_TYPE_OBJECT, true, $resourceClass), $resourceClass)) {
+            if (\is_array($itemConfiguration)) {
+                $fieldConfiguration = array_merge($fieldConfiguration, $itemConfiguration);
```
```suggestion
                $fieldConfiguration += $itemConfiguration;
```

lukasluecke replied:
This does not result in the same behaviour, in case of keys existing in both arrays it would take those of the `$fieldConfiguration` (the "generated" ones). I think it should be able to override them using `$itemConfiguration`, which is why I chose `array_merge`.

alanpoulain commented on src/GraphQl/Type/SchemaBuilder.php:
```diff
@@ -155,20 +173,30 @@ private function getNodeQueryField(): array
-            $queryFields[lcfirst(Inflector::pluralize($shortName))] = $fieldConfiguration;
+        if ($collectionConfiguration && $fieldConfiguration = $this->getResourceFieldConfiguration($resourceClass, $resourceMetadata, null, $deprecationReason, new Type(Type::BUILTIN_TYPE_OBJECT, false, null, true, null, new Type(Type::BUILTIN_TYPE_OBJECT, false, $resourceClass)), $resourceClass)) {
+            if (\is_array($collectionConfiguration)) {
+                $fieldConfiguration = array_merge($fieldConfiguration, $collectionConfiguration);
```
```suggestion
                $fieldConfiguration += $collectionConfiguration;
```

lukasluecke replied:
This does not result in the same behaviour, in case of keys existing in both arrays it would take those of the `$fieldConfiguration` (the "generated" ones). I think it should be able to override them using `$collectionConfiguration`, which is why I chose `array_merge`.

lukasluecke commented:
I will try to get some work done on tests soon, but the rest of the week will be quite busy for me, so not sure when I'll get to it. Thanks for your feedback so far! 🥇

SCIF commented:
Hi @lukasluecke , my message is not about resource/operation definition way but about DX. I tried to suggest a way simpler DX flow of implementation of custom operations. I think that normalization and composing an edge information should not have place in custom operation executor.

lukasluecke commented:
@alanpoulain I have added a simple behat test for `item_query`, do you think it's enough if I add a similar one for `collection_query` for now?

I am still not quite content with the configuration format (`item_query`, `collection_query`) and would prefer a cleaner approach, but I think that can be discussed later.

SCIF commented:
Hi @lukasluecke ,

It would be nice to have tests with parameters.

alanpoulain commented:
@SCIF It is planned in another PR :slightly_smiling_face: This is a first step.

lukasluecke commented:
Thank you very much @alanpoulain 🚀

SCIF commented:
Hi @lukasluecke , @alanpoulain , are you going to implement custom mutations in separate PR?

lukasluecke commented:
There is still #2447 for custom mutations, but I‘ll probably work on mutations and more options in a new PR (if it‘s not merged by the time I get to it).

SCIF commented:
@lukasluecke , @alanpoulain 
But guys, what do you think on idea to extract Normalizing stage out of custom operations implementation?  In REST it's out of custom operation action, which is very convenient. Moreover, it will bring mismatch of behavior between REST and GraphQL in case if you will leave current behaviour.

alanpoulain commented:
@lukasluecke I will finish the PR for custom mutations soon since the major part of the work has already been done.
I have also worked on a way to modify easily the type system.
@SCIF I don't understand what you mean, could you elaborate?

alanpoulain commented:
Great addition, thank you @lukasluecke!

SCIF commented:
@alanpoulain 
Current PR implements custom resolver as a way to execute/invoke custom operation. That means a developer MUST handle normalizing WITHIN every custom operation.

While framework's [REST implementation](https://api-platform.com/docs/core/operations/#recommended-method) handles data type/resource and the developer MUST NOT care on normalizing.

I've tried to express this in [comment above](https://github.com/api-platform/core/pull/2445#issuecomment-464431474).

SCIF commented:
I'm sure, that ability even to implement normalizing stage could be useful, while these are really not common cases.

alanpoulain commented:
OK thank you for the explanation. It could have been done this way if we reused the existing resolvers. But it would have meant less flexibility. I will try to think about it.

SCIF commented:
@alanpoulain ,
So, to be clear, it's much better to call this PR's feature as `custom resolvers` rather than `custom  operation` to distinct different DX.
